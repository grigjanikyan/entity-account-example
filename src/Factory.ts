import { AccountId, AccountRecord } from '@S_E_C_R_E_T/types';
import Modularity, { TAnyValue } from '@S_E_C_R_E_T/modularity';

import { AccountDal } from './Dal';
import { AccountService } from './Service';
import { AccountTransform } from './Transform';
import { AccountController } from './Controller';
import { AccountValidator } from './Validator';
import { AccountValidation } from './Validation';

export * from './Dal';
export { AccountController } from './Controller';

export function accountFactory(
  params: TAnyValue,
): Modularity<
  AccountDal<AccountId, AccountRecord>,
  AccountService<AccountId, AccountRecord>,
  AccountTransform<AccountId, AccountRecord>,
  AccountController<AccountId, AccountRecord>,
  AccountValidator<AccountId, AccountRecord>,
  AccountValidation<AccountId, AccountRecord>
> {
  return new Modularity<
    AccountDal<AccountId, AccountRecord>,
    AccountService<AccountId, AccountRecord>,
    AccountTransform<AccountId, AccountRecord>,
    AccountController<AccountId, AccountRecord>,
    AccountValidator<AccountId, AccountRecord>,
    AccountValidation<AccountId, AccountRecord>
  >({
    params: {
      tableName: 'account',
      fields: [
        'id',
        'version',
        'createdAt',
        'createdBy',
        'updatedAt',
        'updatedBy',
        'removedAt',
        'removedBy',
        'email',
        'password',
        'active',
        'firstName',
        'lastName',
        'mobilePhone',
        'roles',
      ],
      ...params,
    },
    userLayers: {
      Dal: AccountDal,
      Service: AccountService,
      Transform: AccountTransform,
      Controller: AccountController,
      Validator: AccountValidator,
      Validation: AccountValidation,
    },
  });
}
