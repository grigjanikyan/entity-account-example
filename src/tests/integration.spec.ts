import Redis from 'ioredis';
import { SessionJWT, SessionJWTPayload, OtpGeneratePassword } from '@S_E_C_R_E_T/types';
import { IUserSessionRequest, IPromiseResponse } from '@S_E_C_R_E_T/middlewares';
import { S_E_C_R_E_TMailer } from '@S_E_C_R_E_T/email';
import { S_E_C_R_E_TSMS } from '@S_E_C_R_E_T/sms';

import { PG as PgStub } from '@S_E_C_R_E_T/db-pg';
import { Redis as RedisStub } from '@S_E_C_R_E_T/db-redis';

import { otpFactory } from '@S_E_C_R_E_T/entity-otp';
import { accountFactory as factory } from '../Factory';

describe('Account', () => {
  it('updatePassword', async () => {
    const password =
      '73f4a33d231345c656d72cdcbeceaa7543ab9f1e06c87a07d1fc83d21f579c00183fb3cffec3604e328331250695901707e6459a6db42562ba86bb20247b0663';
    const results = [[{ password }], [{ id: 3 }]];

    const queryRunner = new PgStub(results);
    const accountFactory = factory({ queryRunner, getCurrentDateTime: () => 'sommee1etime' });
    let promise = Promise.resolve();
    // eslint-disable-next-line no-return-assign
    const res = ({ promise: jest.fn((prom) => (promise = prom)) } as unknown) as IPromiseResponse;
    accountFactory.controller.updatePassword(
      {
        body: { oldPassword: 'test12345', newPassword: 'test123456' },
        getUser: () => ({ id: 3 }),
      } as IUserSessionRequest<SessionJWTPayload, SessionJWT>,
      res,
    );

    const serviceResult = await promise;
    expect(queryRunner.queries).toMatchSnapshot();
    expect(res.promise).toBeCalled();
    expect(serviceResult).toStrictEqual({ id: 3 });
  });

  it('sendNewPasswordEmail', async () => {
    const results = [[{ password: 'user_pass' }]];
    const queryRunner = new PgStub(results);
    const mailer = {
      restorePassword: jest.fn(),
    };
    const jwt = {
      createRestorePasswordToken: jest.fn(() => "actually it's a hashed password inside JWT"),
    };
    const accountFactory = factory({
      queryRunner,
      mailer: (mailer as unknown) as S_E_C_R_E_TMailer,
      jwt,
    });
    let promise = Promise.resolve();
    // eslint-disable-next-line no-return-assign
    const res = ({ promise: jest.fn((prom) => (promise = prom)) } as unknown) as IPromiseResponse;
    accountFactory.controller.sendNewPasswordEmail(
      {
        body: { email: ' emAiL@domAhhhin.cOm ' },
        getUser: () => ({ id: 3 }),
      } as IUserSessionRequest<SessionJWTPayload, SessionJWT>,
      res,
    );

    const serviceResult = await promise;
    expect(queryRunner.queries).toMatchSnapshot();
    expect(res.promise).toBeCalled();
    expect(mailer.restorePassword).toBeCalledWith(
      { email: 'email@domahhhin.com' },
      {
        password: "actually it's a hashed password inside JWT",
      },
    );
    expect(serviceResult).toStrictEqual(null);
  });

  it('signUp', async () => {
    const results = [[{}], [{}], [{ id: 1112 }]];
    const queryRunner = new PgStub(results);
    const mailer = {
      confirmEmail: jest.fn(),
    };
    const jwt = {
      createEmailToken: jest.fn(() => "actually it's a hashed email inside JWT"),
    };
    const accountFactory = factory({
      queryRunner,
      mailer: (mailer as unknown) as S_E_C_R_E_TMailer,
      jwt,
      getCurrentDateTime: () => 'somme2eetime',
    });
    let promise = Promise.resolve();
    // eslint-disable-next-line no-return-assign
    const res = ({ promise: jest.fn((prom) => (promise = prom)) } as unknown) as IPromiseResponse;
    accountFactory.controller.createUser(
      {
        body: {
          firstName: ' firstName ',
          lastName: ' lastName ',
          email: ' emAiL@domAin.cOm ',
          password: ' superPassword1234+ ',
          mobilePhone: ' +380971707070 ',
        },
        getUser: () => ({ id: 3 }),
      } as IUserSessionRequest<SessionJWTPayload, SessionJWT>,
      res,
    );

    const serviceResult = await promise;
    expect(queryRunner.queries).toMatchSnapshot();
    expect(res.promise).toBeCalled();
    expect(mailer.confirmEmail).toBeCalledWith(
      { email: 'email@domain.com', firstName: 'firstName' },
      { token: "actually it's a hashed email inside JWT" },
    );
    expect(serviceResult).toStrictEqual({ id: 1112 });
  });

  it('createActivationOtp', async () => {
    const redis = new RedisStub([null, 'OK']);
    const sendSms = jest.fn();
    const generatePassword = jest.fn(() => '12345');
    const otp = otpFactory({
      db: (redis as unknown) as Redis.Redis,
      sms: ({ otp: sendSms } as unknown) as S_E_C_R_E_TSMS,
      generatePassword,
    });
    const results = [
      [{ count: '1' }],
      [{ id: 768, mobilePhone: 'phone', mobilePhoneConfirmed: false }],
      [{ id: 768, mobilePhone: 'phone' }],
    ];
    const queryRunner = new PgStub(results);
    const accountFactory = factory({ queryRunner, otpService: otp.service });
    let promise = Promise.resolve();
    // eslint-disable-next-line no-return-assign
    const res = ({ promise: jest.fn((prom) => (promise = prom)) } as unknown) as IPromiseResponse;
    accountFactory.controller.createActivationOtp(
      {
        body: { id: 768 },
        getUser: () => ({ id: 3 }),
      } as IUserSessionRequest<SessionJWTPayload, SessionJWT>,
      res,
    );

    const serviceResult = await promise;
    expect(queryRunner.queries).toMatchSnapshot();
    expect(redis.queries).toMatchSnapshot();
    expect(res.promise).toBeCalled();
    expect(sendSms).toBeCalledWith({ code: '12345', phoneNumber: 'phone' });
    expect(generatePassword).toBeCalledWith({
      length: 5,
      lowercase: false,
      numbers: true,
      symbols: false,
      uppercase: false,
    });
    expect(serviceResult).toStrictEqual({ canRepeatAfterSeconds: 60 });
  });

  it('confirmActivationOtp', async () => {
    const redis = new RedisStub(['89768', 1]);
    const otp = otpFactory({
      db: (redis as unknown) as Redis.Redis,
      sms: ({} as unknown) as S_E_C_R_E_TSMS,
      generatePassword: ({} as unknown) as OtpGeneratePassword,
    });
    const results = [[{ count: '1' }], [{ id: 4356 }]];
    const queryRunner = new PgStub(results);
    const accountFactory = factory({ queryRunner, otpService: otp.service, getCurrentDateTime: () => 'somme3etime' });
    let promise = Promise.resolve();
    // eslint-disable-next-line no-return-assign
    const res = ({ promise: jest.fn((prom) => (promise = prom)) } as unknown) as IPromiseResponse;
    accountFactory.controller.confirmActivationOtp(
      {
        body: { id: 4356, otp: 89768 },
        getUser: () => ({ id: 3 }),
      } as IUserSessionRequest<SessionJWTPayload, SessionJWT>,
      res,
    );
    const serviceResult = await promise;
    expect(queryRunner.queries).toMatchSnapshot();
    expect(redis.queries).toMatchSnapshot();
    expect(res.promise).toBeCalled();
    expect(serviceResult).toStrictEqual({ id: 4356 });
  });

  it('restorePasswordByToken', async () => {
    const results = [[{ id: 4356, email: 'some email' }], [{ id: 4356 }]];
    const queryRunner = new PgStub(results);
    const jwt = {
      verifyRestorePasswordToken: jest.fn(() => ({ email: 'emaila', password: 'passworda' })),
    };
    const mailer = {
      passwordChanged: jest.fn(),
    };
    const accountFactory = factory({ queryRunner, jwt, mailer, getCurrentDateTime: () => 'sommee4etime' });
    let promise = Promise.resolve();
    // eslint-disable-next-line no-return-assign
    const res = ({ promise: jest.fn((prom) => (promise = prom)) } as unknown) as IPromiseResponse;
    accountFactory.controller.restorePasswordByToken(
      {
        body: { token: 'ksjdhfgkf', password: 'test12345+' },
        getUser: () => ({ id: 3 }),
      } as IUserSessionRequest<SessionJWTPayload, SessionJWT>,
      res,
    );
    const serviceResult = await promise;
    expect(queryRunner.queries).toMatchSnapshot();
    expect(res.promise).toBeCalled();
    expect(serviceResult).toStrictEqual(null);
  });

  it('confirmEmailByToken', async () => {
    const queryRunner = new PgStub([[{ id: 876, emailConfirmed: false }], [{ id: 876 }]]);
    const jwt = {
      verifyEmailToken: jest.fn(() => ({ id: 876 })),
    };
    const res = ({ redirect: jest.fn() } as unknown) as IPromiseResponse;
    const next = jest.fn();
    const accountFactory = factory({ queryRunner, jwt, getCurrentDateTime: () => 'sommee4etime123' });
    await accountFactory.controller.confirmEmailByToken(
      ({
        query: { token: 'ksj3434534dhfgkf' },
        getUser: () => ({ id: 3 }),
      } as unknown) as IUserSessionRequest<SessionJWTPayload, SessionJWT>,
      res,
      next,
    );
    expect(queryRunner.queries).toMatchSnapshot();
    expect(next).toBeCalledTimes(0);
    expect(res.redirect).toBeCalledWith(302, 'http://localhost:8080/email-confirmed-successfully');
  });

  describe('update user', () => {
    it('email changed', async () => {
      const results = [
        [
          {
            id: 66666,
            email: 'email@dom1.com',
            firstName: 'firstNadfgme',
            lastName: 'lastdsfName',
            mobilePhone: '+380971705670',
          },
        ],
        [{ id: 66666 }],
      ];
      const queryRunner = new PgStub(results);
      const mailer = {
        confirmEmail: jest.fn(),
      };
      const jwt = {
        createEmailToken: jest.fn(() => "actually it's a hashed email inside JWT123123123123"),
      };
      const accountFactory = factory({
        queryRunner,
        mailer: (mailer as unknown) as S_E_C_R_E_TMailer,
        jwt,
        getCurrentDateTime: () => 'sommee5etime',
      });
      let promise = Promise.resolve();
      // eslint-disable-next-line no-return-assign
      const res = ({ promise: jest.fn((prom) => (promise = prom)) } as unknown) as IPromiseResponse;
      accountFactory.controller.update(
        ({
          params: {
            id: '66666',
          },
          body: {
            firstName: ' firstNadfgme ',
            lastName: ' lastdsfName ',
            email: ' emAiL@dom2.cOm ',
            password: ' superPsword1234+ ',
            mobilePhone: ' +380971705670 ',
          },
          getUser: () => ({ id: 356 }),
        } as unknown) as IUserSessionRequest<SessionJWTPayload, SessionJWT>,
        res,
      );

      const serviceResult = await promise;
      expect(queryRunner.queries).toMatchSnapshot();
      expect(res.promise).toBeCalled();
      expect(mailer.confirmEmail).toBeCalledWith(
        { email: 'email@dom2.com', firstName: 'firstNadfgme' },
        { token: "actually it's a hashed email inside JWT123123123123" },
      );
      expect(serviceResult).toStrictEqual({ id: 66666 });
    });

    it('email same', async () => {
      const results = [
        [
          {
            id: 66666,
            email: 'email@dom1.com',
            firstName: 'firstNadfgme',
            lastName: 'lastdsfName',
            mobilePhone: '+380971705670',
          },
        ],
        [{ id: 66666 }],
      ];
      const queryRunner = new PgStub(results);
      const mailer = {
        confirmEmail: jest.fn(),
      };
      const jwt = {
        createEmailToken: jest.fn(() => "actually it's a hashed email inside JWT123123123123"),
      };
      const accountFactory = factory({
        queryRunner,
        mailer: (mailer as unknown) as S_E_C_R_E_TMailer,
        jwt,
        getCurrentDateTime: () => 'sommeee6time',
      });
      let promise = Promise.resolve();
      // eslint-disable-next-line no-return-assign
      const res = ({ promise: jest.fn((prom) => (promise = prom)) } as unknown) as IPromiseResponse;
      accountFactory.controller.update(
        ({
          params: {
            id: '66666',
          },
          body: {
            firstName: ' firstNadfgme ',
            lastName: ' lastdsfName ',
            email: ' emAiL@dom1.cOm ',
            password: ' superPsword1234+ ',
            mobilePhone: ' +380971705670 ',
          },
          getUser: () => ({ id: 356 }),
        } as unknown) as IUserSessionRequest<SessionJWTPayload, SessionJWT>,
        res,
      );

      const serviceResult = await promise;
      expect(queryRunner.queries).toMatchSnapshot();
      expect(res.promise).toBeCalled();
      expect(mailer.confirmEmail).toBeCalledTimes(0);
      expect(serviceResult).toStrictEqual({ id: 66666 });
    });
  });
});
