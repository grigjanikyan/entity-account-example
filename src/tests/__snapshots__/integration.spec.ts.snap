// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account confirmActivationOtp 1`] = `
Array [
  Object {
    "bindings": Array [
      4356,
    ],
    "sql": "select count(\\"id\\") from \\"account\\" where \\"id\\" in ($1)",
  },
  Object {
    "bindings": Array [
      1,
      "somme3etime",
      4356,
      4356,
    ],
    "sql": "update \\"account\\" set \\"version\\" = $1, \\"updated_at\\" = $2, \\"updated_by\\" = $3 where \\"id\\" = $4 returning \\"id\\"",
  },
]
`;

exports[`Account confirmActivationOtp 2`] = `
Array [
  Object {
    "args": Object {
      "key": "account-activation-otp-4356",
    },
    "method": "get",
  },
  Object {
    "args": Object {
      "key": "account-activation-otp-4356",
    },
    "method": "del",
  },
]
`;

exports[`Account confirmEmailByToken 1`] = `
Array [
  Object {
    "bindings": Array [
      876,
      false,
    ],
    "sql": "select \\"id\\", \\"version\\", \\"created_at\\", \\"created_by\\", \\"updated_at\\", \\"updated_by\\", \\"removed_at\\", \\"removed_by\\", \\"email\\", \\"password\\", \\"active\\", \\"first_name\\", \\"last_name\\", \\"mobile_phone\\", \\"roles\\" from \\"account\\" where \\"id\\" = $1 and \\"email_confirmed\\" = $2",
  },
  Object {
    "bindings": Array [
      1,
      "sommee4etime123",
      3,
      876,
    ],
    "sql": "update \\"account\\" set \\"version\\" = $1, \\"updated_at\\" = $2, \\"updated_by\\" = $3 where \\"id\\" = $4 returning \\"id\\"",
  },
]
`;

exports[`Account createActivationOtp 1`] = `
Array [
  Object {
    "bindings": Array [
      768,
    ],
    "sql": "select count(\\"id\\") from \\"account\\" where \\"id\\" in ($1)",
  },
  Object {
    "bindings": Array [
      768,
    ],
    "sql": "select \\"id\\", \\"version\\", \\"created_at\\", \\"created_by\\", \\"updated_at\\", \\"updated_by\\", \\"removed_at\\", \\"removed_by\\", \\"email\\", \\"password\\", \\"active\\", \\"first_name\\", \\"last_name\\", \\"mobile_phone\\", \\"roles\\" from \\"account\\" where \\"id\\" = $1",
  },
  Object {
    "bindings": Array [
      768,
    ],
    "sql": "select \\"id\\", \\"email\\", \\"first_name\\", \\"last_name\\", \\"mobile_phone\\" from \\"account\\" where \\"id\\" = $1",
  },
]
`;

exports[`Account createActivationOtp 2`] = `
Array [
  Object {
    "args": Object {
      "key": "account-activation-otp-768",
    },
    "method": "get",
  },
  Object {
    "args": Object {
      "key": "account-activation-otp-768",
      "ttl": 60,
      "val": "12345",
    },
    "method": "setex",
  },
]
`;

exports[`Account restorePasswordByToken 1`] = `
Array [
  Object {
    "bindings": Array [
      "emaila",
      "passworda",
    ],
    "sql": "select \\"id\\", \\"version\\", \\"created_at\\", \\"created_by\\", \\"updated_at\\", \\"updated_by\\", \\"removed_at\\", \\"removed_by\\", \\"email\\", \\"password\\", \\"active\\", \\"first_name\\", \\"last_name\\", \\"mobile_phone\\", \\"roles\\" from \\"account\\" where \\"email\\" = $1 and \\"password\\" = $2",
  },
  Object {
    "bindings": Array [
      1,
      "sommee4etime",
      3,
      "8fe5b36117dc7c5a3862f86741665ac31171cf9380f866af84c68da2eb346ff3b5dcfbc76eea8f3fe5cbda96a6e4b022e5bd421e5cc529a8940aaecab0ea6f96",
      4356,
    ],
    "sql": "update \\"account\\" set \\"version\\" = $1, \\"updated_at\\" = $2, \\"updated_by\\" = $3, \\"password\\" = $4 where \\"id\\" = $5 returning \\"id\\"",
  },
]
`;

exports[`Account sendNewPasswordEmail 1`] = `
Array [
  Object {
    "bindings": Array [
      "email@domahhhin.com",
    ],
    "sql": "select \\"id\\", \\"version\\", \\"created_at\\", \\"created_by\\", \\"updated_at\\", \\"updated_by\\", \\"removed_at\\", \\"removed_by\\", \\"email\\", \\"password\\", \\"active\\", \\"first_name\\", \\"last_name\\", \\"mobile_phone\\", \\"roles\\" from \\"account\\" where \\"email\\" = $1",
  },
]
`;

exports[`Account signUp 1`] = `
Array [
  Object {
    "bindings": Array [
      "email@domain.com",
    ],
    "sql": "select count(\\"email\\") from \\"account\\" where \\"email\\" in ($1)",
  },
  Object {
    "bindings": Array [
      "+380971707070",
    ],
    "sql": "select count(\\"mobile_phone\\") from \\"account\\" where \\"mobile_phone\\" in ($1)",
  },
  Object {
    "bindings": Array [
      "somme2eetime",
      3,
      "email@domain.com",
      "firstName",
      "lastName",
      "+380971707070",
      "50e9c5e18dbebbd9f3f6ee44aaa6f4df6545fa445f6aede5e76a589498787a942b21e104b919d820114a012fb46b711e16d31334650c180b1b1b5fe907a16106",
      null,
      null,
      0,
    ],
    "sql": "insert into \\"account\\" (\\"created_at\\", \\"created_by\\", \\"email\\", \\"first_name\\", \\"last_name\\", \\"mobile_phone\\", \\"password\\", \\"removed_at\\", \\"removed_by\\", \\"version\\") values ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10) returning \\"id\\"",
  },
]
`;

exports[`Account update user email changed 1`] = `
Array [
  Object {
    "bindings": Array [
      356,
    ],
    "sql": "select \\"id\\", \\"version\\", \\"created_at\\", \\"created_by\\", \\"updated_at\\", \\"updated_by\\", \\"removed_at\\", \\"removed_by\\", \\"email\\", \\"password\\", \\"active\\", \\"first_name\\", \\"last_name\\", \\"mobile_phone\\", \\"roles\\" from \\"account\\" where \\"id\\" = $1",
  },
  Object {
    "bindings": Array [
      1,
      "sommee5etime",
      356,
      "email@dom2.com",
      "firstNadfgme",
      "lastdsfName",
      "+380971705670",
      356,
    ],
    "sql": "update \\"account\\" set \\"version\\" = $1, \\"updated_at\\" = $2, \\"updated_by\\" = $3, \\"email\\" = $4, \\"first_name\\" = $5, \\"last_name\\" = $6, \\"mobile_phone\\" = $7 where \\"id\\" = $8 returning \\"id\\"",
  },
]
`;

exports[`Account update user email same 1`] = `
Array [
  Object {
    "bindings": Array [
      356,
    ],
    "sql": "select \\"id\\", \\"version\\", \\"created_at\\", \\"created_by\\", \\"updated_at\\", \\"updated_by\\", \\"removed_at\\", \\"removed_by\\", \\"email\\", \\"password\\", \\"active\\", \\"first_name\\", \\"last_name\\", \\"mobile_phone\\", \\"roles\\" from \\"account\\" where \\"id\\" = $1",
  },
  Object {
    "bindings": Array [
      1,
      "sommeee6time",
      356,
      "email@dom1.com",
      "firstNadfgme",
      "lastdsfName",
      "+380971705670",
      356,
    ],
    "sql": "update \\"account\\" set \\"version\\" = $1, \\"updated_at\\" = $2, \\"updated_by\\" = $3, \\"email\\" = $4, \\"first_name\\" = $5, \\"last_name\\" = $6, \\"mobile_phone\\" = $7 where \\"id\\" = $8 returning \\"id\\"",
  },
]
`;

exports[`Account updatePassword 1`] = `
Array [
  Object {
    "bindings": Array [
      3,
      "e6e8ba72bf2bab68c923599a08489c6f3a35018e870d490fa25b4c2a2d82361093b9a8f1d0cdd02e5a7dd5e714dc090263e2e90af3d3d861c056f28a4adb2d95",
    ],
    "sql": "select \\"id\\", \\"version\\", \\"created_at\\", \\"created_by\\", \\"updated_at\\", \\"updated_by\\", \\"removed_at\\", \\"removed_by\\", \\"email\\", \\"password\\", \\"active\\", \\"first_name\\", \\"last_name\\", \\"mobile_phone\\", \\"roles\\" from \\"account\\" where \\"id\\" = $1 and \\"password\\" = $2",
  },
  Object {
    "bindings": Array [
      1,
      "sommee1etime",
      3,
      "73f4a33d231345c656d72cdcbeceaa7543ab9f1e06c87a07d1fc83d21f579c00183fb3cffec3604e328331250695901707e6459a6db42562ba86bb20247b0663",
      3,
    ],
    "sql": "update \\"account\\" set \\"version\\" = $1, \\"updated_at\\" = $2, \\"updated_by\\" = $3, \\"password\\" = $4 where \\"id\\" = $5 returning \\"id\\"",
  },
]
`;
